### PR TITLE
[spaceship] adopt to new API protocol for 2FA

### LINKED ISSUE #7556: Lane gets cancel at 2 step verification.
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description
I have a line written, which sends latest dsyms files to crashlytics. This works fine unitl now, that I update to latest fastlane. The same is if I try to use `lane beta`.  It always fails when it comes to 2 step verification.

### Captured Output

Command Used: `refresh_latest_dsyms`
<details><summary>Output/Log</summary>

```

[14:48:42]: -------------------------------------------------
[14:48:42]: --- Step: Verifying required fastlane version ---
[14:48:42]: -------------------------------------------------
[14:48:42]: Your fastlane version 2.0.5 matches the minimum requirement of 1.110.0  ✅
[14:48:42]: ------------------------------
[14:48:42]: --- Step: default_platform ---
[14:48:42]: ------------------------------
[14:48:43]: Driving the lane 'ios refresh_latest_dsyms' 🚀
[14:48:43]: -----------------------
[14:48:43]: --- Step: cocoapods ---
[14:48:43]: -----------------------
[14:48:43]: $ bundle exec pod install
[14:48:46]: ▸ Analyzing dependencies
[14:48:46]: ▸ Downloading dependencies
[14:48:46]: ▸ Using AHKNavigationController (1.3)
[14:48:46]: ▸ Using Bolts (1.8.4)
[14:48:46]: ▸ Using Crashlytics (3.8.3)
[14:48:46]: ▸ Using DGRunkeeperSwitch (1.1.4)
[14:48:46]: ▸ Using FBSDKCoreKit (4.18.0)
[14:48:46]: ▸ Using FBSDKLoginKit (4.18.0)
[14:48:46]: ▸ Using Fabric (1.6.11)
[14:48:46]: ▸ Using KYShutterButton (2.0.0)
[14:48:46]: ▸ Using MZTimerLabel (0.5.4)
[14:48:46]: ▸ Using Parse (1.14.2)
[14:48:46]: ▸ Using ParseFacebookUtilsV4 (1.11.1)
[14:48:46]: ▸ Using SDiPhoneVersion (1.1.2)
[14:48:46]: ▸ Using Zip (0.6.0)
[14:48:46]: ▸ Using iOSDFULibrary (3.0.0)
[14:48:46]: ▸ Generating Pods project
[14:48:47]: ▸ Integrating client project
[14:48:49]: ▸ Sending stats
[14:48:49]: ▸ Pod installation complete! There are 11 dependencies from the Podfile and 14 total pods installed.
[14:48:49]: ------------------------------
[14:48:49]: --- Step: get_build_number ---
[14:48:49]: ------------------------------
[14:48:49]: $ cd projectPath && agvtool what-version -terse
[14:48:50]: ▸ 103
[14:48:50]: ----------------------------
[14:48:50]: --- Step: download_dsyms ---
[14:48:50]: ----------------------------
[14:48:50]: Login to iTunes Connect (my email)
+------------------+--------------------------+
|                Lane Context                 |
+------------------+--------------------------+
| DEFAULT_PLATFORM | ios                      |
| PLATFORM_NAME    | ios                      |
| LANE_NAME        | ios refresh_latest_dsyms |
| BUILD_NUMBER     | 103                      |
+------------------+--------------------------+
[14:48:52]: Invalid 2 step response {"trustedPhoneNumbers"=>[{"id"=>1, "verified"=>true, "pending"=>false, "lastTwoDigits"=>"04", "countryCode"=>"SI", "countryDialCode"=>"386", "createDate"=>"09/21/2016 11:49:56 AM", "updateDate"=>"09/21/2016 11:49:56 AM", "number"=>"******04", "pushMode"=>"sms", "obfuscatedNumber"=>"••••••04", "rawNumber"=>"******04", "numberWithDialCode"=>"+386 ••••••04"}], "securityCode"=>{"length"=>6, "tooManyCodesSent"=>false, "tooManyCodesValidated"=>false, "securityCodeLocked"=>false}, "authenticationType"=>"hsa2", "recoveryUrl"=>"https://iforgot.apple.com/phone/add?prs_account_nm=myEmail&autoSubmitAccount=true&appId=142", "cantUsePhoneNumberUrl"=>"https://iforgot.apple.com/iforgot/phone/add?context=cantuse&prs_account_nm=myEmail&autoSubmitAccount=true&appId=142", "recoveryWebUrl"=>"https://iforgot.apple.com/password/verify/appleid?prs_account_nm=myEmail&autoSubmitAccount=true&appId=142", "repairPhoneNumberUrl"=>"https://gsa.apple.com/appleid/account/manage/repair/verify/phone", "repairPhoneNumberWebUrl"=>"https://appleid.apple.com/widget/account/?#!repair", "aboutTwoFactorAuthenticationUrl"=>"https://support.apple.com/kb/HT204921", "hsa2Account"=>true, "trustedPhoneNumber"=>{"id"=>1, "verified"=>true, "pending"=>false, "lastTwoDigits"=>"04", "countryCode"=>"SI", "countryDialCode"=>"386", "createDate"=>"09/21/2016 11:49:56 AM", "updateDate"=>"09/21/2016 11:49:56 AM", "number"=>"******04", "pushMode"=>"sms", "obfuscatedNumber"=>"••••••04", "rawNumber"=>"******04", "numberWithDialCode"=>"+386 ••••••04"}, "supportsRecovery"=>true, "managedAccount"=>false}

+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | Verifying required fastlane version | 0           |
| 2    | default_platform                    | 0           |
| 3    | cocoapods                           | 6           |
| 4    | get_build_number                    | 1           |
| 💥   | download_dsyms                      | 2           |
+------+-------------------------------------+-------------+

[14:48:52]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...



```

</details>

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                 | Value                                       |
| ------------------- | ------------------------------------------- |
| OS                  | 10.12.2                                     |
| Ruby                | 2.3.0                                       |
| Bundler?            | true                                        |
| Git                 | git version 2.10.1 (Apple Git-78)           |
| Installation Source | ~/.rvm/gems/ruby-2.3.0/bin/fastlane         |
| Host                | Mac OS X 10.12.2 (16C67)                    |
| Ruby Lib Dir        | ~/.rvm/rubies/ruby-2.3.0/lib                |
| OpenSSL Version     | OpenSSL 1.0.2i  22 Sep 2016                 |
| Is contained        | false                                       |
| Xcode Path          | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version       | 8.2                                         |

### System Locale

| Error                       |
| --------------------------- |
| No Locale with UTF8 found 🚫 |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Customise this file, documentation can be found here:
# https://github.com/fastlane/fastlane/tree/master/fastlane/docs
# All available actions: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Actions.md
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
fastlane_version "1.110.0"

def get_display_name
  get_info_plist_value(path: "./baseball/Info.plist", key: "CFBundleDisplayName")
end

default_platform :ios

platform :ios do
  before_all do
    ENV["SLACK_URL"] = "mySlackWebhookURL"
    cocoapods
  end

  desc "Runs all the tests"
  lane :test do
    scan
  end

  desc "Submit a new Beta Build to Apple TestFlight"
  desc "This will also make sure the profile is up to date"


  lane :beta do |options|
    ensure_git_status_clean
    initialBranch = git_branch
    sh "git checkout betaRelease"
    sh "git merge --no-ff develop"
    increment_build_number
    gitMessage = options[:whatsNew] || ""
    commit_version_bump(message: "Increment build number to #{get_build_number}.\n\n#{gitMessage}")
    
    tag = get_version_number
    if !git_tag_exists(tag: tag)
	    add_git_tag(tag: tag)
    end
    push_git_tags
    sh "git checkout " + initialBranch
    
    gym(scheme: "baseball",
        include_bitcode: true)
    testflight
    refresh_latest_dsyms
    slackMessage = options[:whatsNew] || "No description :)"
    slack(channel: "#general",
	  default_payloads: [],
	  payload: {
            "Built by" => "Klemen",
            "What\'s new" => slackMessage
          },
          message: "Successfully released beta " + get_display_name + " Baseball v" +get_version_number + " (" + get_build_number + ")")
  end

  lane :betaPrematureFail do |options|
  slackMessage = options[:whatsNew] || "No description :)"
  slack(channel: "#general",
    default_payloads: [],
	  payload: {
            "Built by" => "Klemen",
            "What\'s new" => slackMessage
          },
    message: "Successfully released beta " + get_display_name + " Baseball v" + get_version_number + " (" + get_build_number + ")")
end

  desc "Deploy a new version to the App Store"
  lane :release do
    # match(type: "appstore")
    # snapshot
    gym(scheme: "baseball") # Build your app - more options available
    deliver(force: true)
    # frameit
  end

  lane :refresh_dsyms do
    download_dsyms                  # Download dSYM files from iTC
    upload_symbols_to_crashlytics   # Upload them to Crashlytics
    clean_build_artifacts           # Delete the local dSYM files
  end

  lane :refresh_latest_dsyms do
    buildNumber = get_build_number
    download_dsyms(build_number: buildNumber)# Download dSYM files from iTC
    upload_symbols_to_crashlytics        # Upload them to Crashlytics
    clean_build_artifacts                # Delete the local dSYM files
  end

  after_all do |lane|
    # This block is called, only if the executed lane was successful

    # slack(
    #   message: "Successfully deployed new App Update."
    # )
  end

  error do |lane, exception|
    # slack(
    #   message: exception.message,
    #   success: false
    # )
  end
end

lane :refresh_dsyms do
  download_dsyms                  # Download dSYM files from iTC
  upload_symbols_to_crashlytics   # Upload them to Crashlytics
  clean_build_artifacts           # Delete the local dSYM files
end


# More information about multiple platforms in fastlane: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Platforms.md
# All available actions: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Actions.md

# fastlane reports which actions are used
# No personal data is recorded. Learn more at https://github.com/fastlane/enhancer
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier "appIdentifier" # The bundle identifier of your app
apple_id "myEmail" # Your Apple email address

team_id "myTeamID"  # Developer Portal Team ID

# you can even provide different app identifiers, Apple IDs and team names per lane:
# More information: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md


itc_team_id "myTeamID"```
</details>

### fastlane gems

| Gem                 | Version | Update-Status |
| ------------------- | ------- | ------------- |
| credentials_manager | 0.16.2  | ✅ Up-To-Date  |
| spaceship           | 0.39.0  | ✅ Up-To-Date  |
| fastlane            | 2.0.5   | ✅ Up-To-Date  |


### Loaded fastlane plugins:

**No plugins Loaded**


<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.0.0        |
| executable-hooks          | 1.3.2        |
| bundler-unload            | 1.0.2        |
| rubygems-bundler          | 1.4.4        |
| bundler                   | 1.13.6       |
| io-console                | 0.4.5        |
| CFPropertyList            | 2.3.4        |
| i18n                      | 0.7.0        |
| json                      | 1.8.3        |
| minitest                  | 5.10.1       |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| activesupport             | 4.2.7.1      |
| public_suffix             | 2.0.4        |
| addressable               | 2.5.0        |
| babosa                    | 1.0.2        |
| claide                    | 1.0.1        |
| fuzzy_match               | 2.0.4        |
| nap                       | 1.1.0        |
| cocoapods-core            | 1.1.1        |
| cocoapods-deintegrate     | 1.0.1        |
| cocoapods-downloader      | 1.1.2        |
| cocoapods-plugins         | 1.0.0        |
| cocoapods-search          | 1.0.0        |
| cocoapods-stats           | 1.0.0        |
| netrc                     | 0.7.8        |
| cocoapods-trunk           | 1.1.1        |
| cocoapods-try             | 1.1.0        |
| colored                   | 1.2          |
| escape                    | 0.0.4        |
| fourflusher               | 2.0.1        |
| gh_inspector              | 1.0.2        |
| molinillo                 | 0.5.4        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.1        |
| cocoapods                 | 1.1.1        |
| highline                  | 1.7.8        |
| commander                 | 4.4.1        |
| security                  | 0.1.3        |
| credentials_manager       | 0.16.2       |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| dotenv                    | 2.1.1        |
| excon                     | 0.54.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.10.0       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.10.1       |
| fastimage                 | 2.0.1        |
| jwt                       | 1.5.6        |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.1.0        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| hurley                    | 0.2          |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| google-api-client         | 0.9.20       |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.2.0        |
| rubyzip                   | 1.2.0        |
| slack-notifier            | 1.5.1        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.2        |
| terminal-table            | 1.7.3        |
| word_wrap                 | 1.0.0        |
| xcode-install             | 2.0.9        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| xcpretty-travis-formatter | 0.0.4        |
</details>


*generated on:* **2016-12-17**
</details>

### PR DESCRIPTION
**Pull request by hjanuschka**
API seems to have changed.

should fix: https://github.com/fastlane/fastlane/issues/7556

### CODE REVIEW

File: spaceship/lib/spaceship/two_step_client.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def handle_two_step(response)
         result = choose(*available)
         device_id = result.match(/.*\t.*\t\((.*)\)/)[1]
         select_device(r, device_id)
-      elsif r.body.kind_of?(Hash) && r.body["phoneNumberVerification"].kind_of?(Hash)
+      elsif r.body.kind_of?(Hash) && r.body["trustedPhoneNumbers"].kind_of?(Hash)
```
Comment (meyskens) [inline spaceship/lib/spaceship/two_step_client.rb]: `trustedPhoneNumbers` is an array, so this check will fail. In my test code `r.body["trustedPhoneNumbers"][0].kind_of?(Hash)` worked so an extra check on the array and the first element may be nice (soon all elements? not sure how Apple will change this)

**Review by meyskens — CHANGES_REQUESTED** [review]

File: spaceship/lib/spaceship/two_step_client.rb (hunk 1 above)
Comment (hjanuschka) [inline spaceship/lib/spaceship/two_step_client.rb]: @meyskens thx for the feedback, i did a commit to handle the array, can you re-test it?

**Review by meyskens — APPROVED** [review]
That should work

**Comment by hjanuschka** [comment]
@meyskens can you run it and verify it?

**Comment by meyskens** [comment]
Works fine now 😄

**Comment by hjanuschka** [comment]
@meyskens awesome, thx for your help!

File: spaceship/lib/spaceship/two_step_client.rb [hunk 2]
```diff
@@ -35,7 +35,7 @@ def handle_two_step(response)
         result = choose(*available)
         device_id = result.match(/.*\t.*\t\((.*)\)/)[1]
         select_device(r, device_id)
-      elsif r.body.kind_of?(Hash) && r.body["phoneNumberVerification"].kind_of?(Hash)
+      elsif r.body.kind_of?(Hash) && r.body["trustedPhoneNumbers"].kind_of?(Array) && r.body["trustedPhoneNumbers"].first.kind_of?(Hash)
```
Comment (mpirri) [inline spaceship/lib/spaceship/two_step_client.rb]: What is the importance of checking `"trustedPhoneNumbers"` here? It doesn't appear that we ever use the value, or the hash it contains.

Also, should we be checking the existence/type of `response.body["securityCode"]`, which is now used on line 51 below?

Comment (meyskens) [inline spaceship/lib/spaceship/two_step_client.rb]: Apple has changed the response to have multiple phone numbers listed for 2FA. See #7562

Comment (mpirri) [inline spaceship/lib/spaceship/two_step_client.rb]: But these values in the body are never actually used, right?  So couldn't we just delete these tests here, or else update the test to check `r.body["securityCode"]` in order to support line 51 below?  (previously this line was testing the response to see if line 51 could execute properly)

Comment (meyskens) [inline spaceship/lib/spaceship/two_step_client.rb]: To answer that I don't know enough about the internals of Fastlane.

**Comment by akaralar** [comment]
I want to test this, how do I do that? I followed the steps in toolsanddebugging.md but somehow I get errors when I try `bundle update` in my project root folder

File: spaceship/lib/spaceship/two_step_client.rb (hunk 2 above)
Comment (hjanuschka) [inline spaceship/lib/spaceship/two_step_client.rb]: @mpirri true point - i just tried to fix it, so should we just check on  `securityCode` 
i am just not sure if this is enough - can we have some more 👀  on this one?

**Comment by KrauseFx** [comment]
@akaral You didn't tell us what kind of errors

File: spaceship/lib/spaceship/two_step_client.rb (hunk 2 above)
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: We still use `phoneNumberVerification` in the `handle_two_factor` method, shouldn't we check for this?

Comment (hjanuschka) [inline spaceship/lib/spaceship/two_step_client.rb]: @KrauseFx `phoneNumberVerification` is not used anymore, its changed in favor of: `trustedPhoneNumbers`

**Review by hjanuschka — COMMENTED** [review]
.

**Comment by akaralar** [comment]
@KrauseFx sorry I was in a hurry to upload a build, cheated fastlane and uploaded with Xcode in the end :D To summarize my problem:

- I clone the fastlane library with `git clone`
- I cd into the cloned folder, run `bundle install`, finishes successfully
- I run `bundle exec rspec` to make sure everything is working correctly, no erros no issues.
- I copy the `Gemfile` from `.assets` folder in the cloned repo, and paste it to my Xcode project root folder.
- I change the `local_fastlane_path` to be `local_fastlane_path = "~/Development/github/fastlane"`, which points to the cloned directory.
- I run `bundle update` and get the error: `The path `/Users/akaralar/Development/github/fastlane_core` does not exist.`
It shouldn't exists at this location since it is looking for `fastlane_core` 1 level above the repo's root directory.

- If I change the `local_fastlane_path` to point to the `fastlane` directory inside the cloned directory, (ie `local_fastlane_path = "~/Development/github/fastlane/fastlane"`), then I get the error: `The path `/Users/akaralar/Development/github/fastlane/fastlane/fastlane_core` does not exist.`

I hope that clears up things a bit, sorry for not elaborating on this earlier.

**Review by mpirri — APPROVED** [review]

**Pull request merged**